File: week4/cpp/bugeyeImage.cpp
/*
 Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
 
 This program is distributed WITHOUT ANY WARRANTY to the
 Plus and Premium membership students of the online course
 titled "Computer Visionfor Faces" by Satya Mallick for
 personal non-commercial use.
 
 Sharing this code is strictly prohibited without written
 permission from Big Vision LLC.
 
 For licensing and other inquiries, please email
 spmallick@bigvisionllc.com
 
 */

#include <iostream>
#include <dlib/opencv.h>
#include <opencv2/highgui/highgui.hpp>
#include <dlib/image_processing/frontal_face_detector.h>
#include <dlib/image_processing.h>
#include <opencv2/core/core.hpp>
#include <opencv2/imgproc/imgproc.hpp>

using namespace cv;
using namespace std;
using namespace dlib;

#define FACE_DOWNSAMPLE_RATIO_DLIB 1    //affects dlib's face detector

#ifndef M_PI
  #define M_PI 3.14159
#endif

Mat barrel(Mat &src, float k)
{
  int w= src.cols;
  int h= src.rows;
  
  // Meshgrid of destiation image
  Mat Xd = cv::Mat::zeros(src.size(), CV_32F);
  Mat Yd = cv::Mat::zeros(src.size(), CV_32F);
  
  float Xu, Yu;
  for (int y = 0; y < h; y++)
  {
    for (int x = 0; x < w; x++)
    {
      Xu = ( (float) x / w )- 0.5;
      Yu = ( (float) y / h )- 0.5;
      
      float r = sqrt(Xu * Xu + Yu * Yu);

      // Pincushion distortion function
      float rn = std::min( (double)r, r + (pow(r,k) - r) * cos(M_PI * r ));
      
      // Applying the distortion on the grid
      Xd.at<float>(y,x) =  w * (rn * Xu / r + 0.5);
      Yd.at<float>(y,x) = h * (rn * Yu / r + 0.5);
    }
  }

  // Interpolation of points
  Mat dst;
  cv::remap( src, dst, Xd, Yd, CV_INTER_CUBIC, BORDER_CONSTANT, Scalar(0,0, 0) );
  return dst;
}


int main(int argc, char** argv)
{
  frontal_face_detector detector = get_frontal_face_detector();
  shape_predictor pose_model;

  string modelPath = "../../common/shape_predictor_68_face_landmarks.dat";

  string filename = "../data/images/sample.jpg";

  float bulge_amount = 2;

  int radius = 30;

  // accept command line arguments for model path for landmark detector and image file
  cout << "USAGE" << endl << "./bugeyeImage <bulge_amount default : 2 > < radius around eye default : 30 > <filename> " << endl;

  if (argc == 2)
  {   
    bulge_amount = atof(argv[1]);
  }
  else if(argc == 3)
  {
    bulge_amount = atof(argv[1]);
    radius = atoi(argv[2]);
  }
  else if(argc == 4)
  {
    bulge_amount = atof(argv[1]);
    radius = atoi(argv[2]);
    filename = argv[3];
  }

  deserialize(modelPath) >> pose_model;

  // load a nice picture
  double t = (double)cv::getTickCount();

  Mat src = imread(filename);

  cv_image<bgr_pixel> cimg(src);
  std::vector<dlib::rectangle> faces;

  faces = detector(cimg);

  // Find the pose of each face.
  full_object_detection landmarks;

  // Find the landmark points using DLIB Facial landmarks detector
  landmarks = pose_model(cimg, faces[0]);

  // Find the roi for left and right Eye
  Rect roiEyeRight ( (landmarks.part(43).x()-radius)
                    , (landmarks.part(43).y()-radius)
                    , ( landmarks.part(46).x() - landmarks.part(43).x() + 2*radius )
                    , ( landmarks.part(47).y() - landmarks.part(43).y() + 2*radius ) );
  Rect roiEyeLeft ( (landmarks.part(37).x()-radius)
                   , (landmarks.part(37).y()-radius)
                   , ( landmarks.part(40).x() - landmarks.part(37).x() + 2*radius )
                   , ( landmarks.part(41).y() - landmarks.part(37).y() + 2*radius ) );

  // Find the patch and apply the transformation
  Mat eyeRegion, output;
  output = src.clone();
  src(roiEyeRight).copyTo(eyeRegion);
  eyeRegion = barrel(eyeRegion, bulge_amount);
  eyeRegion.copyTo(output(roiEyeRight));
  src(roiEyeLeft).copyTo(eyeRegion);
  eyeRegion = barrel(eyeRegion, bulge_amount);
  eyeRegion.copyTo(output(roiEyeLeft));

  cout << "time taken " << ((double)cv::getTickCount() - t)/cv::getTickFrequency() << endl;
  // imshow("distorted",dst);
  imshow("final",output);

  imwrite("results/bugeye.jpg",output);
  waitKey(0);
     
  return 0;
}

File: week4/python/bugeyeImage.py
#!/usr/bin/python
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
#
# This code is made available to the students of
# the online course titled "Computer Vision for Faces"
# by Satya Mallick for personal non-commercial use.
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC.
#
# For licensing and other inquiries, please email
# spmallick@bigvisionllc.com
#
import cv2,dlib,time,argparse
import numpy as np

FACE_DOWNSAMPLE_RATIO = 1
modelPath = "../../common/shape_predictor_68_face_landmarks.dat"
detector = dlib.get_frontal_face_detector()
predictor = dlib.shape_predictor(modelPath)
filename = '../data/images/auriane2.jpg'
bulge_amount = 3
radius = 30
print ("USAGE : python bugeyeImage.py -f file.jpg -b bulge amount ( default : 2 ) -r radius (radius around eye, default : 30 )" )

def barrel(src, k):
  w = src.shape[1]
  h = src.shape[0]

  # Meshgrid of destiation image
  Xu, Yu = np.meshgrid(np.arange(w), np.arange(h))

  Xu = np.float32(Xu)/w - 0.5
  Yu = np.float32(Yu)/h - 0.5

  XuSquare = np.square(Xu)
  YuSquare = np.square(Yu)

  r = np.sqrt(XuSquare + YuSquare)

  # Pincushion distortion function
  rn = np.minimum(r, r + np.multiply((np.power(r,k) - r ) , np.cos(np.pi * r) ))

  # Applying the distortion on the grid
  Xd = w * ( cv2.divide( np.multiply( rn, Xu ), r) + 0.5 )
  Yd = h * ( cv2.divide( np.multiply( rn, Yu ), r) + 0.5 )

  # Interpolation of points
  dst = cv2.remap(src, Xd, Yd, cv2.INTER_CUBIC)
  return dst


def getLandmarks(im):
  imSmall = cv2.resize(im,None,
                       fx=1.0/FACE_DOWNSAMPLE_RATIO,
                       fy=1.0/FACE_DOWNSAMPLE_RATIO,
                       interpolation = cv2.INTER_LINEAR)
  #detect faces
  rects = detector(imSmall, 0)
  if len(rects) == 0:
    return 1

  #scale the points before sending to the pose predictor as we will send the original image
  newRect = dlib.rectangle(int(rects[0].left()*FACE_DOWNSAMPLE_RATIO),
                           int(rects[0].top()*FACE_DOWNSAMPLE_RATIO),
                           int(rects[0].right()*FACE_DOWNSAMPLE_RATIO),
                           int(rects[0].bottom()*FACE_DOWNSAMPLE_RATIO))
  points = []
  [points.append((p.x, p.y)) for p in predictor(im, newRect).parts()]
  return points


ap = argparse.ArgumentParser()
ap.add_argument("-f", "--filename", help="Path to the image")
ap.add_argument("-r", "--radius",  help="radius around eye, default : 30")
ap.add_argument("-b", "--bulge",  help="bulge amount default : 2")

args = vars(ap.parse_args())

if(args["filename"]):
  filename = args["filename"]
if(args["radius"]):
  radius = args["radius"]
if(args["bulge"]):
  bulge_amount = float(args["bulge"])

src = cv2.imread(filename)
# Find the landmark points using DLIB Facial landmarks detector
landmarks = getLandmarks(src)

# Find the roi for left Eye
roiEyeLeft = [ landmarks[37][0] - radius, landmarks[37][1] - radius,
          (landmarks[40][0] - landmarks[37][0] + 2*radius),
          (landmarks[41][1] - landmarks[37][1] + 2*radius)  ]

# Find the roi for right Eye
roiEyeRight = [ landmarks[43][0] - radius, landmarks[43][1] - radius,
          (landmarks[46][0] - landmarks[43][0] + 2*radius),
          (landmarks[47][1] - landmarks[43][1] + 2*radius)  ]

output = np.copy(src)
# Find the patch for left eye and apply the transformation
eyeRegion = src[roiEyeLeft[1]:roiEyeLeft[1] + roiEyeLeft[3],roiEyeLeft[0]:roiEyeLeft[0] + roiEyeLeft[2]]
eyeRegion = barrel(eyeRegion, bulge_amount);
output[roiEyeLeft[1]:roiEyeLeft[1] + roiEyeLeft[3],roiEyeLeft[0]:roiEyeLeft[0] + roiEyeLeft[2]] = eyeRegion

# Find the patch for right eye and apply the transformation
eyeRegion = src[roiEyeRight[1]:roiEyeRight[1] + roiEyeRight[3],roiEyeRight[0]:roiEyeRight[0] + roiEyeRight[2]]
eyeRegion = barrel(eyeRegion, bulge_amount);
output[roiEyeRight[1]:roiEyeRight[1] + roiEyeRight[3],roiEyeRight[0]:roiEyeRight[0] + roiEyeRight[2]] = eyeRegion

cv2.imshow('distorted',output)
cv2.imwrite('results/bugeye.jpg',output)
cv2.waitKey(0)
